pytealutils/application/client.py

```diff
@@ -0,0 +1,66 @@
+from os import urandom
+from typing import List
+
+from algosdk.abi import Contract, Method
+from algosdk.account import address_from_private_key
+from algosdk.atomic_transaction_composer import (
+    AtomicTransactionComposer,
+    TransactionSigner,
+)
+from algosdk.v2client.algod import AlgodClient
+
+# TODO: Cache suggested params
+
+
+class ContractClient:
+    def __init__(
+        self, client: AlgodClient, contract: Contract, signer: TransactionSigner = None
+    ):
+        self.client = client
+        self.networks = contract.networks
+        self.contract = contract
+        self.signer = signer
+
+        self.addr = address_from_private_key(self.signer.private_key)
+
+        for m in contract.methods:
+            print(m.name)
+            setattr(self, m.name, lambda args, budget=1: self.call(m, args, budget))
+
```

Comment:
Maybe this will fix the scoping issue? (Inspired by my [previous successful attempt](https://github.com/algorand/go-algorand/pull/3277/files?authenticity_token=plRQEIbruDaHZj1woDLwofeXs2nIncj7Evo5WmGxXHZQ1WeEFiymIm_TvqIQei8dB5tlRD8HAa3L6xZKVjEGJA&file-filters%5B%5D=&w=1#diff-40701f29ae43a257ecb1a5d7e954c250367078cd6741c88f4c83000d23848cf6R305))

```suggestion
        for m in contract.methods:
            print(m.name)
            caller = self._get_caller(m, args, budget)
            setattr(self, m.name, caller)
    def _get_caller(self, m, args, budget):
        def call(args, budget=1):
            self.call(m, args, budget)
        return call
```